server.py

```diff
@@ -1,62 +1,84 @@
-from mcrcon import MCRcon
-import re, json, asyncio, websockets, os, signal, sys, logging
+import asyncio
+import json
+import logging
+import os
+import signal
+from dataclasses import asdict
 
-RCON_HOST = os.getenv('RCON_HOST')
-RCON_PASSWORD = os.getenv('RCON_PASSWORD')
-RCON_PORT = os.getenv('RCON_PORT', default=25575)
-SOCKET_HOST = os.getenv('SOCKET_HOST', default='0.0.0.0')
-SOCKET_PORT = int(os.getenv('SOCKET_PORT', default=8888))
-LOG_LEVEL = os.getenv('LOG_LEVEL', default='ERROR')
-REFRESH_RATE = int(os.getenv('REFRESH_RATE', default=5))
+import websockets
 
-logging.basicConfig(level=LOG_LEVEL)
+from mclocations.minecraft import MinecraftServer
 
-assert RCON_HOST is not None
-assert RCON_PASSWORD is not None
+log = logging.getLogger(__name__)
+
+
+class WebSocketServer:
+    def __init__(self, minecraft, socket_host, socket_port, refresh_rate):
+        self.log = log
+        self.minecraft = minecraft
+        self.socket_host = socket_host
+        self.socket_port = socket_port
+        self.refresh_rate = refresh_rate
+        self.ws_server = None
+
+    async def start(self):
+        self.log.info('Starting the websocket server')
+
+        def stop_cb():
+            return asyncio.create_task(self.stop())
 
-async def sendLocations(websocket, path):
-    while True:
-        output = json.dumps(getLocations())
         try:
-            await websocket.send(output)
-        except (websockets.exceptions.ConnectionClosedOK, websockets.exceptions.ConnectionClosedError, websockets.exceptions.ConnectionClosed):
-            await websocket.wait_closed()
-            logging.info("Client connection closed")
-            break
-        finally:
-            await asyncio.sleep(REFRESH_RATE)
+            asyncio.get_running_loop().add_signal_handler(signal.SIGINT, stop_cb)
+            asyncio.get_running_loop().add_signal_handler(signal.SIGTERM, stop_cb)
+        except NotImplementedError as e:
+            log.exception(f'Not implemented')
+
+        self.ws_server = await websockets.serve(self.send_locations, self.socket_host, self.socket_port)
+
+        await self.ws_server.wait_closed()
+
+        self.log.info('Websocket server closed')
+
+    async def stop(self):
+        self.ws_server.close()
+        self.minecraft.disconnect()
 
-def getLocations():
-    output = []
-    response = mcr.command("/list")
-    match = re.search(r"([0-9]+).+:(.+)", response)
-    numberOfPlayers = int(match.group(1))
+    async def send_locations(self, websocket, path):
+        while True:
+            players = self.minecraft.get_players()
```

Comment:
As we talked, this is where each connected web client would be issuing rcon commands to the minecraft server.

Reply:
Yep. I think that could be split out and try to use a queue or something more fun to communicate between tasks.